### PR TITLE
Implement GROUP_CONCAT() function

### PR DESCRIPTION
**Pull request by nrother**
As per cakephp/cakephp#9529.

On SQLServer this seems to be impossible to implement, as [a subquery is neccessary](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/f09d4166-2030-41fe-b86e-392fbc94db53/tsql-equivalent-for-groupconcat-function?forum=transactsql).

It's currently not possible to pass in a `ExpressionInterface`, to achive something like `GROUP_CONCAT(id+1), but it looks like that's not supported for the other functions either. (And I have no idea how to implement it).

_I don't have tested the generated queries against a real DB server_ I can test against MySQL later, but I don't have access to the other types of servers.

I'm still slightly confused by the Database system, so please tell me if there is a better way to achive this ;)

### CODE REVIEW

File: src/Database/Dialect/PostgresDialectTrait.php [hunk 1]
```diff
@@ -107,6 +107,18 @@ protected function _transformFunctionExpression(FunctionExpression $expression)
                 // CONCAT function is expressed as exp1 || exp2
                 $expression->name('')->tieWith(' ||');
                 break;
+            case 'GROUP_CONCAT':
+                $expression
+                    ->name('array_to_string')
+                    ->iterateParts(function ($p, $key) use ($expression) {
+                        if ($key === 0) {
+                            return new FunctionExpression('array_agg', [$p => 'literal'], [], 'string');
+                        } else {
```
Comment (dereuromark) [inline src/Database/Dialect/PostgresDialectTrait.php]: no else necessary.

File: src/Database/FunctionsBuilder.php [hunk 2]
```diff
@@ -136,6 +138,22 @@ public function concat($args, $types = [])
     {
         return $this->_build('CONCAT', $args, $types, 'string');
     }
+    
+    /**
+     * Returns a FunctionExpression representing a group concatenation.
+     *
+     * @param mixed $expression the function argument
+     * @param string|null $separator Separator to use
```
Comment (dereuromark) [inline src/Database/FunctionsBuilder.php]: looks like just string type here.

File: tests/TestCase/Database/Driver/SqliteTest.php [hunk 3]
```diff
@@ -143,4 +144,37 @@ public function testSchemaValue($input, $expected)
...
     }
+    
+    /**
+     * Tests that GROUP_CONCAT is transformed correctly
+     *
+     * @return void
+     */
+    public function testGroupConcatTransform()
+    {
+        $driver = $this->getMockBuilder('Cake\Database\Driver\Sqlite')
```
Comment (dereuromark) [inline tests/TestCase/Database/Driver/SqliteTest.php]: Can we use ::class syntax here since this is PHP5.5+?

Comment (nrother) [inline tests/TestCase/Database/Driver/SqliteTest.php]: The rest of the file is using string syntax also, might be something for a more general cleanup.,,

File: src/Database/FunctionsBuilder.php [hunk 4]
```diff
@@ -142,7 +142,7 @@ public function concat($args, $types = [])
     /**
      * Returns a FunctionExpression representing a group concatenation.
      *
-     * @param mixed $expression the function argument
+     * @param string $expression the function argument
```
Comment (dereuromark) [inline src/Database/FunctionsBuilder.php]: I meant the line below (separator), see my previous comment

**Comment by nrother** [comment]
I'm not quite sure why the tests are failing, seems unrelated to me.

File: tests/TestCase/Database/Driver/SqliteTest.php [hunk 5]
```diff
@@ -143,4 +144,37 @@ public function testSchemaValue($input, $expected)
...
+            ->setMethods(['_connect'])
+            ->getMock();
+        $connection = $this
+            ->getMockBuilder('\Cake\Database\Connection')
+            ->setMethods(['connect'])
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $query = new \Cake\Database\Query($connection);
```
Comment (markstory) [inline tests/TestCase/Database/Driver/SqliteTest.php]: You could add a `use` statement for this class.

**Comment by markstory** [comment]
Tests are failing on [travis ci](https://travis-ci.org/cakephp/cakephp/jobs/163552482#L691) because your new SQLServer test isn't skipping itself when the extension is not installed.

**Comment by lorenzo** [comment]
The need for a subquery does not make this feature impossible to implement in SQL Server, we do that in many other places

**Comment by nrother** [comment]
@lorenzo Do you have a pointer where something is done with a subquery? The problem is not the subquery itself, but the need to bind it properly to the outer query. For me this looks very complicated, given that the outer query might not select the primary key of the table etc.

**Comment by lorenzo** [comment]
@nrother there are a few ways. This is one of them: https://github.com/cakephp/cakephp/commit/a0bf28d2dfa6da22bdfd1365cb4bc3fb4563b2ea#diff-52ed7bc997253a0b2000d3d676601fd8R87

There is also a suggestion in the link you provided, where the group concat is done with a `CTE` expression, maybe that's a possibility

**Comment by lorenzo** [comment]
@nrother this is another example https://github.com/cakephp/cakephp/blob/master/src/Database/Dialect/SqlserverDialectTrait.php#L145-L188

**Comment by nrother** [comment]
Ok, this is really getting complicated. I somehow figured out how it could be done, but I'm still not completely convinced that this transformation is possible in any case (joins etc.).

My current sketch is here: https://gist.github.com/nrother/abd9f9499ecfe7a21db25e6409ccd2ae

Currently I'm fighting with the builder to get the rather ... arcane subquery in the select clause (see gist) done. *I'm more than happy for a helping hand by someone more experienced/any pointers!*

At the moment I'd rather drop support for SqlServer for GROUP_CONCAT, probably it will always be a huge hack.

**Comment by nrother** [comment]
Another solution I came across would be the usage of a user defined function that implements GROUP_CONCAT, [as described here](https://www.simple-talk.com/sql/t-sql-programming/concatenating-row-values-in-transact-sql/). But this would requires us to install this function prior to each query, and probably cause privilege problems. Don't like this at the moment.

The third option in to just require the use to install [an extension for SQLServer that is implementing GROUP_CONCAT](http://groupconcat.codeplex.com/)...

**Comment by nrother** [comment]
A forth option might be to remove the GROUP_CONCAT() operator from the query and implement it as a map/reduce step. This would obviously a step way from a pure DB abstraction layer.

To summaries:
- Throw `LogicException`
  - :+1: Simple
  - :-1: No support for SqlServer
- Implement with `FOR XML PATH`
  - 👍 The way it should be done
  - 👎 Really complicated, probably impossible in edge cases
- Implement with UDF/require an extension
  - 👍 Easy, full support
  - 👎 Requires user interaction/changes to the DB system
- Implement in PHP as map/reduce step
  - 👍 Full support, somewhat easy to implement
  - 👎 Might not be what the user expect

Votes/opinions welcome. I'd currently go for option 1, with the possibility to implement 2 or 4 later.

BTW: The queries for MySQL, Sqlite and Postgres have be verified by me now.

**Comment by lorenzo** [comment]
If we cannot provide support for all drivers, then why add the function? You can already make the function call yourself without having to add it to the API.

I vote for exploring the subquery options a bit more. Requiring the extension can also be a good way to solve this, but that would be the last option

**Comment by nrother** [comment]
I just pushed a implementation that should work for Sqlserver. The code looks a bit... complicated, abusing query modifiers as "tags", not sure if there is something better available, I'm open for any tips 😄 

This probably needs some fine-tuning, just want to check out if the general idea is good. Bad thing is, that queries selecting more than the one `GROUP_CONCAT` column will not work with this, and implementing this is probably tough.

TODO:
* Don't change column name to "expr"
* Think about multi-colum queries
* Fix the option to use native group_concat if installed, add tests for that

**Comment by markstory** [comment]
Closing as this can no longer be merged. If you have time to get it into a mergeable state please open a new pull request :smile:

**Pull request closed without merging**